internal/graph/check.go

```diff
@@ -952,6 +960,185 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+		}
+		usersetName, relation := tuple.SplitObjectRelation(t.GetUser())
+		if relation == "" {
+			// This should not be possible given we have put in the type restriction.
+			return nil, fmt.Errorf("expect user %s to have form objection relation", t.GetUser())
+		}
+		if info.userToUsersetMapping.Exists(usersetName) {
+			return &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolveCheckResponseMetadata{
+					DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + info.dsCount.Load(),
+				},
+			}, nil
+		}
+		usersetItems = append(usersetItems, usersetName)
+	}
+
+	// when we get here, it means that we are not able to find intersection with the current sortedUsersetSlices.
+	// We will recursively check whether recursiveMatchUserUserset to find out whether there are intersections.
+	// TODO: parallelize the lookup.
+	for _, user := range usersetItems {
+		_, visited := info.visitedUserset.LoadOrStore(user, struct{}{})
+		if !visited {
+			newReq := req.clone()
+			newReq.TupleKey.Object = user
+			result, err := recursiveMatchUserUserset(ctx, newReq, info)
+			if err != nil {
+				return nil, err
+			}
+			if result.GetAllowed() {
+				return result, nil
+			}
+		}
+		// Note that visited does not necessary means that there are cycles.  For the following model,
+		// type user
+		// type group
+		//   relations
+		//     define member: [user, group#member]
+		// We have something like
+		// group:1#member@group:2#member
+		// group:1#member@group:3#member
+		// group:2#member@group:a#member
+		// group:3#member@group:a#member
+		// Note that both group:2#member and group:3#member has group:a#member. However, they are not cycles.
+	}
+
+	// There are no intersections and recursive matching shows there are no intersections.
+	return &ResolveCheckResponse{
+		Allowed: false,
+		ResolutionMetadata: &ResolveCheckResponseMetadata{
+			DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + info.dsCount.Load(),
+		},
+	}, nil
+}
+
+func nestedUsersetFastpath(ctx context.Context,
+	typesys *typesystem.TypeSystem,
+	ds storage.RelationshipTupleReader,
+	req *ResolveCheckRequest) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "nestedUsersetFastpath")
+	defer span.End()
+
+	// Note that this will take care of the publicly assignable case as well.
+	iter, err := checkutil.IteratorReadStartingFromUser(ctx,
+		typesys, ds, req,
+		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
+		nil)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+	filteredIter := storage.NewConditionsFilteredTupleKeyIterator(
+		storage.NewFilteredTupleKeyIterator(
+			storage.NewTupleKeyIteratorFromTupleIterator(iter),
+			validation.FilterInvalidTuples(typesys),
+		),
+		checkutil.BuildTupleKeyConditionFilter(ctx, req.GetContext(), typesys),
+	)
+	defer filteredIter.Stop()
+
+	group, err := storage.GetAllObjects(ctx, filteredIter)
```

Comment:
Why pull in all objects when you can be scanning the results as they are yielded?